feat(aws): add support for geoproximity routing

prasadkatti opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

This PR adds support for [Geoproximity routing](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy-geoproximity.html) in route53 to the AWS provider.

It adds four new annotations to enable this support.

***Testing done***

1. I created a hosted zone called `example.test`
2. Applied the manifests below to my cluster
3. Ran a single control loop of external-dns locally
4. Verified on the AWS console that new records for the services were created
5. Deleted the services
6. Ran another loop of external-dns
7. Verified the records were deleted from Route53.

Also did some more testing along these lines to ensure there are no unnecessary Upserts when the service isn't changed.

I used these service manifests for testing - 

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-test-region
  annotations:
    external-dns.alpha.kubernetes.io/hostname: nginx-test.example.test
    external-dns.alpha.kubernetes.io/target: 8.8.8.8
    external-dns.alpha.kubernetes.io/set-identifier: region
    external-dns.alpha.kubernetes.io/aws-geoproximitylocation-aws-region: us-west-2
    external-dns.alpha.kubernetes.io/aws-geoproximitylocation-bias: "10"
spec:
  ports:
  - port: 80
    targetPort: 80
  selector:
    app: nginx

---

apiVersion: v1
kind: Service
metadata:
  name: nginx-test-coordinates
  annotations:
    external-dns.alpha.kubernetes.io/hostname: nginx-test.example.test
    external-dns.alpha.kubernetes.io/target: 8.8.8.8
    external-dns.alpha.kubernetes.io/set-identifier: coordinates
    external-dns.alpha.kubernetes.io/aws-geoproximitylocation-coordinates: 90,90
spec:
  ports:
  - port: 80
    targetPort: 80
  selector:
    app: nginx

---

# I didn't test this one, but it should work

# apiVersion: v1
# kind: Service
# metadata:
#   name: nginx-test-local-zone
#   annotations:
#     external-dns.alpha.kubernetes.io/hostname: nginx-test.example.test
#     external-dns.alpha.kubernetes.io/target: 8.8.8.8
#     external-dns.alpha.kubernetes.io/set-identifier: localzone
#     external-dns.alpha.kubernetes.io/aws-geoproximitylocation-local-zone-group: "usw2-pdx1-az1"
# spec:
#   ports:
#   - port: 80
#     targetPort: 80
#   selector:
#     app: nginx
```

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #4927 

**Checklist**

- [x] Unit tests updated
- [ ] End user documentation updated.  Haven't yet updated the docs. Will get to it once my others questions get resolved.

prasadkatti commented on provider/aws/aws.go:
```diff
@@ -53,16 +53,20 @@ const (
+	providerSpecificGeolocationContinentCode           = "aws/geolocation-continent-code"
+	providerSpecificGeolocationCountryCode             = "aws/geolocation-country-code"
+	providerSpecificGeolocationSubdivisionCode         = "aws/geolocation-subdivision-code"
+	providerSpecificGeoProximityLocationAWSRegion      = "aws/geoproximitylocation-aws-region"
+	providerSpecificGeoProximityLocationBias           = "aws/geoproximitylocation-bias"
+	providerSpecificGeoProximityLocationCoordinates    = "aws/geoproximitylocation-coordinates"
+	providerSpecificGeoProximityLocationLocalZoneGroup = "aws/geoproximitylocation-local-zone-group"
```
These are the 4 new annotations that will be added. Rest of the change is just whitespaces and alignment.

prasadkatti commented on provider/aws/aws.go:
```diff
@@ -832,6 +850,23 @@ func (p *AWSProvider) AdjustEndpoints(endpoints []*endpoint.Endpoint) ([]*endpoi
+		// Add provider specific properties to the endpoint if record has set identifier
+		if ep.SetIdentifier != "" {
+
+			// check if ep has any of the geoproximity properties
+			_, ok1 := ep.GetProviderSpecificProperty(providerSpecificGeoProximityLocationAWSRegion)
+			_, ok2 := ep.GetProviderSpecificProperty(providerSpecificGeoProximityLocationLocalZoneGroup)
+			_, ok3 := ep.GetProviderSpecificProperty(providerSpecificGeoProximityLocationCoordinates)
+
+			if ok1 || ok2 || ok3 {
+				// check if ep has bias property
+				if _, ok := ep.GetProviderSpecificProperty(providerSpecificGeoProximityLocationBias); !ok {
+					// if not, set it to 0
+					ep.SetProviderSpecificProperty(providerSpecificGeoProximityLocationBias, "0")
+				}
+			}
```
If there is a more idiomatic way to express this, please let me know. I am not yet very experienced with Go.

prasadkatti replied:
But basically what is happening here is, when a record with geoproximity routing is read from r53, it will always have a value for _Bias_. If the _Bias_ annotation is missing on the source, then the actual endpoint and the source endpoint don't match. That leads to unnecessary Upserts. 
IIUC, the AdjustEndpoints function is used to "normalize" endpoints and avoid such unnecessary changes.

prasadkatti commented on provider/aws/aws_test.go:
```diff
@@ -1902,7 +1956,7 @@ func validateEndpoints(t *testing.T, provider *AWSProvider, endpoints []*endpoin
 
 	normalized, err := provider.AdjustEndpoints(endpoints)
 	assert.NoError(t, err)
-	assert.True(t, testutils.SameEndpoints(normalized, expected), "actual and normalized endpoints don't match. %+v:%+v", endpoints, normalized)
+	assert.True(t, testutils.SameEndpoints(normalized, expected), "normalized and expected endpoints don't match. %+v:%+v", normalized, expected)
```
I am having some trouble understanding this. Doesn't this mean that the actual endpoints, the normaized endpoints and the expected endpoints ALL need to match? But how can that work when the normalization process in the AdjustEndpoints method is actually changing the endpoint?

I added a [test case](https://github.com/kubernetes-sigs/external-dns/pull/5347/files#diff-9f91016006712e3748e634b63c1babae4614850350719699a2c8f710534560e4R622-R631) where I don't specify the _Bias_ on the desired endpoint. But that gets normalized and the test fails. What am I doing wrong?

prasadkatti replied:
Ohh maybe I know what is wrong. The records passed into the newAWSProvider func are supposed to be the actual records and not the desired endpoints. So then it should always have the Bias value, since R53 will always have a Bias value on the geoproximity records. So maybe the test case is invalid then.

ivankatliarchuk commented:
/ok-to-test

ivankatliarchuk commented:
For this to review
- shared how it was tested with results
- share kubernetes manifests so that we could validate that as well

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -53,16 +53,20 @@ const (
+	providerSpecificGeolocationContinentCode           = "aws/geolocation-continent-code"
+	providerSpecificGeolocationCountryCode             = "aws/geolocation-country-code"
+	providerSpecificGeolocationSubdivisionCode         = "aws/geolocation-subdivision-code"
+	providerSpecificGeoProximityLocationAWSRegion      = "aws/geoproximitylocation-aws-region"
```
`aws-region` not required, just `region`

prasadkatti replied:
I have tried to match it with the field names in the [GeoProximityLocation](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/route53@v1.51.1/types#GeoProximityLocation) type

ivankatliarchuk replied:
`aws/geoproximitylocation-aws-region`

does not looks right

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -542,6 +546,20 @@ func (p *AWSProvider) records(ctx context.Context, zones map[string]*profiledZon
 									ep.WithProviderSpecific(providerSpecificGeolocationSubdivisionCode, *r.GeoLocation.SubdivisionCode)
 								}
 							}
+						case r.GeoProximityLocation != nil:
```
should be it's own method

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -832,6 +850,23 @@ func (p *AWSProvider) AdjustEndpoints(endpoints []*endpoint.Endpoint) ([]*endpoi
 		}
+
+		// Add provider specific properties to the endpoint if record has set identifier
+		if ep.SetIdentifier != "" {
```
should be it's own method

prasadkatti replied:
done, thanks

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -926,6 +961,39 @@ func (p *AWSProvider) newChange(action route53types.ChangeAction, ep *endpoint.E
 			change.ResourceRecordSet.GeoLocation = geolocation
 		}
+
+		geoproximity := &route53types.GeoProximityLocation{}
```
what if geoproximity not enabled? why we even have it here?

prasadkatti replied:
Honestly, I looked at how Geolocation works and tried to follow suit.

If there are no annotations corresponding to GeoProximity, then the `GeoProximityLocation` field of the `ResourceRecordSet` just stays `nil`. The field is populated iff any of the geoproximity annotations are seen.

Also, I wish there was a separate annotation for specifying routing policy. But since that isn't present, I have to check for the presence of all possible geoproximity annotations before I can definitively say that the record is not using geoproximity.

ivankatliarchuk replied:
This way it could be a breaking change, so unsure

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -926,6 +961,39 @@ func (p *AWSProvider) newChange(action route53types.ChangeAction, ep *endpoint.E
+
+		geoproximity := &route53types.GeoProximityLocation{}
+		useGeoproximity := false
+		if prop, ok := ep.GetProviderSpecificProperty(providerSpecificGeoProximityLocationAWSRegion); ok {
```
Move to own method, do not bloat current pls

prasadkatti replied:
done, thanks

ivankatliarchuk reviewed (changes requested):
pls share code coverage for new code as well

prasadkatti commented:
@ivankatliarchuk - First of all, thanks for the taking the time to review. I have pushed another commit to address some of your review comments. I have also updated the PR description to include the manifests used for testing this. Let me know if you need anything else.

I am still trying to figure out how to get code coverage for just the files I touched. I ran `make cover-html` and that shows me that I have added good coverage for the lines I added. I don't know how to share that with you.

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -560,6 +566,22 @@ func (p *AWSProvider) records(ctx context.Context, zones map[string]*profiledZon
 	return endpoints, nil
 }
 
+func handleGeoProximityLocationRecord(r *route53types.ResourceRecordSet, ep *endpoint.Endpoint) {
```
Here is suggested improvements

1. Use Early Returns: If the GeoProximityLocation is nil, return early to avoid unnecessary checks.
2. Slightly rewrite the logic, to simplify readability.

```
if r.GeoProximityLocation == nil {
		return
	}

	if region := aws.ToString(r.GeoProximityLocation.AWSRegion); region != "" {
		ep.WithProviderSpecific(providerSpecificGeoProximityLocationAWSRegion, region)
	}

	if bias := r.GeoProximityLocation.Bias; bias != nil {
		ep.WithProviderSpecific(providerSpecificGeoProximityLocationBias, fmt.Sprintf("%d", aws.ToInt32(bias)))
	}

	if coords := r.GeoProximityLocation.Coordinates; coords != nil {
		coordinates := fmt.Sprintf("%s,%s", aws.ToString(coords.Latitude), aws.ToString(coords.Longitude))
		ep.WithProviderSpecific(providerSpecificGeoProximityLocationCoordinates, coordinates)
	}

	if localZoneGroup := aws.ToString(r.GeoProximityLocation.LocalZoneGroup); localZoneGroup != "" {
		ep.WithProviderSpecific(providerSpecificGeoProximityLocationLocalZoneGroup, localZoneGroup)
	}
```

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -939,6 +982,40 @@ func (p *AWSProvider) newChange(action route53types.ChangeAction, ep *endpoint.E
 	return change
 }
 
+func buildChangeForGeoProximityEndpoint(change *Route53Change, ep *endpoint.Endpoint) {
```
```suggestion
func witchChangeForGeoProximityEndpoint(change *Route53Change, ep *endpoint.Endpoint) {
```

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -939,6 +982,40 @@ func (p *AWSProvider) newChange(action route53types.ChangeAction, ep *endpoint.E
 }
 
+func buildChangeForGeoProximityEndpoint(change *Route53Change, ep *endpoint.Endpoint) {
+	geoproximity := &route53types.GeoProximityLocation{}
```
This method complexity is quite high, we need to split it most likely, example

```go
func buildChangeForGeoProximityEndpoin(......)
  geoproximity := &route53types.GeoProximityLocation{}
  isGeoproximity := false
  
  isGeoproximity = withGeoProximityAWSRegion(ep, geoproximity) || useGeoproximity
  isGeoproximity = withGeoProximityBias(ep, geoproximity) || useGeoproximity
  isGeoproximity = withGeoProximityCoordinates(ep, geoproximity) || useGeoproximity
  isGeoproximity = withGeoProximityLocalZoneGroup(ep, geoproximity) || useGeoproximity
  
  if isGeoproximity {
    change.ResourceRecordSet.GeoProximityLocation = geoproximity
  }
}


where
func withGeoProximityLocalZoneGroup(ep *endpoint.Endpoint, geoproximity *route53types.GeoProximityLocation) bool {
	if prop, ok := ep.GetProviderSpecificProperty(providerSpecificGeoProximityLocationLocalZoneGroup); ok {
		geoproximity.LocalZoneGroup = aws.String(prop)
		return true
	}
	return false
}
```

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -939,6 +982,40 @@ func (p *AWSProvider) newChange(action route53types.ChangeAction, ep *endpoint.E
+	if prop, ok := ep.GetProviderSpecificProperty(providerSpecificGeoProximityLocationCoordinates); ok {
+		coordinates := strings.Split(prop, ",")
+		if len(coordinates) == 2 {
+			latitude := coordinates[0]
+			longitude := coordinates[1]
+			geoproximity.Coordinates = &route53types.Coordinates{
+				Latitude:  aws.String(latitude),
+				Longitude: aws.String(longitude),
+			}
```
Would be nice to add some validation here as well

```
type Coordinates struct {

	//  Specifies a coordinate of the north–south position of a geographic point on
	// the surface of the Earth (-90 - 90).
	//
	// This member is required.
	Latitude *string

	//  Specifies a coordinate of the east–west position of a geographic point on the
	// surface of the Earth (-180 - 180).
	//
	// This member is required.
	Longitude *string
}
```

Probably something like 
```go
// validateCoordinates checks if the given latitude and longitude are valid.
func validateCoordinates(lat, lon string) error {
	latitude, err := strconv.ParseFloat(lat, 64)
	if err != nil || latitude < -90 || latitude > 90 {
		return errors.New("invalid latitude: must be a number between -90 and 90")
	}

	longitude, err := strconv.ParseFloat(lon, 64)
	if err != nil || longitude < -180 || longitude > 180 {
		return errors.New("invalid longitude: must be a number between -180 and 180")
	}

	return nil
}

// Example usage
func validateAndSetCoordinates(coordinates string) (*route53types.Coordinates, error) {
	parts := strings.Split(coordinates, ",")
	if len(parts) != 2 {
		return nil, errors.New("invalid coordinates format: expected 'latitude,longitude'")
	}

	lat := strings.TrimSpace(parts[0])
	lon := strings.TrimSpace(parts[1])

	if err := validateCoordinates(lat, lon); err != nil {
		return nil, err
	}

	return &route53types.Coordinates{
		Latitude:  aws.String(lat),
		Longitude: aws.String(lon),
	}, nil
}
```

ivankatliarchuk commented on provider/aws/aws.go:
```diff
@@ -939,6 +982,40 @@ func (p *AWSProvider) newChange(action route53types.ChangeAction, ep *endpoint.E
+			}
+			useGeoproximity = true
+		} else {
+			log.Errorf("Invalid coordinates format for %s: %s; expected format 'latitude,longitude'", providerSpecificGeoProximityLocationCoordinates, prop)
```
This is debatable, but most likely should be log.Warn, to warn user so they need to change configuraiton

prasadkatti replied:
I feel an `Errorf` here is okay, because the user is going to expect e-dns to create those records but it won't. That object with the invalid coordinates is just going to fail to reconcile.
We are also using an `Errorf` [here](https://github.com/prasadkatti/external-dns/blob/d587756d062c74af544aaa44696ad08d2b9275b5/provider/aws/aws.go#L896) even when the record will get created.

ivankatliarchuk reviewed (changes requested):
would be nice to add docs with example setup and links to AWS docs.

https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/aws.md

ivankatliarchuk commented:
Any chance you share here a result of a smoke test? With geoproximity and without geoproximity settings, share relevant kube manifests as well pls. Does it require AWS IAM specific actions?

Docs needs updating as well https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/aws.md

ivankatliarchuk commented:
/label tide/merge-method-squash

prasadkatti commented:
New changes -
- updated the docs by adding the new annotations for geoproximity
- added some more tests. The code coverage should be pretty good for the code that was updated.
---

> Does it require AWS IAM specific actions?

No. It doesn't. It's just the same actions except the routing policy will be different.

> Any chance you share here a result of a smoke test? With geoproximity and without geoproximity settings, share relevant kube manifests as well pls.

I have shared a manifest that I have been using for testing in the PR description. Let me know if that doesn't work for some reason. Here is a screenshot of the records that get added for those services in that manifest.
![CleanShot 2025-06-09 at 01 24 31](https://github.com/user-attachments/assets/5fbe39e3-74ac-4c9d-9dbc-0871de1d324b)

Let me know if you need anything else from me at this point. Thanks so much for all your help so far @ivankatliarchuk !

prasadkatti commented on docs/tutorials/aws.md:
```diff
@@ -891,9 +891,14 @@ For any given DNS name, only **one** of the following routing policies can be us
 - Latency-based routing: `external-dns.alpha.kubernetes.io/aws-region`
 - Failover:`external-dns.alpha.kubernetes.io/aws-failover`
 - Geolocation-based routing:
-  - `external-dns.alpha.kubernetes.io/aws-geolocation-continent-code`
```
with 2 spaces, the rendering of the nested list doesn't happen correctly.
![CleanShot 2025-06-09 at 01 43 05](https://github.com/user-attachments/assets/b0284a95-2d41-4d52-8b0b-2ef3b461a175)

prasadkatti replied:
so I updated the indent from 2 spaces to 4 spaces. But that made the linter sad. :(
I will change it back to 2 spaces just to make linter pass.

ivankatliarchuk reviewed (changes requested):
Found few areas that require code coverage as well
<img width="1387" alt="Screenshot 2025-06-09 at 11 30 12" src="https://github.com/user-attachments/assets/f03f9284-89d0-4a00-b0dc-680e34fa5100" />
<img width="1649" alt="Screenshot 2025-06-09 at 11 30 21" src="https://github.com/user-attachments/assets/dd70ce1b-397f-42e7-bb33-2cc0c0b9b8d0" />

ivankatliarchuk reviewed (commented):
/lgtm

cc: @mloiseleur

prasadkatti commented on docs/tutorials/aws.md:
```diff
@@ -894,6 +894,11 @@ For any given DNS name, only **one** of the following routing policies can be us
+  - `external-dns.alpha.kubernetes.io/aws-geoproximitylocation-aws-region`
+  - `external-dns.alpha.kubernetes.io/aws-geoproximitylocation-local-zone-group`
+  - `external-dns.alpha.kubernetes.io/aws-geoproximitylocation-coordinates`
+  - `external-dns.alpha.kubernetes.io/aws-geoproximitylocation-bias`
```
yeah, sounds good to me. Pushed up a change to use the shorter annotations.

prasadkatti replied:
@mloiseleur - I have updated the annotations. Should I mark this as resolved? Let me know if you have any other feedback for the PR that I can work on. Thanks.
